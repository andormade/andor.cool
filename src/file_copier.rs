use std::collections::hash_map::DefaultHasher;
use std::fs::{self, File};
use std::hash::Hasher;
use std::io::{Read};
use std::path::Path;
use crate::error::MyError;

pub fn copy_file_with_versioning(source_path: &str, destination_dir: &str) -> Result<String, MyError> {
    let source_path = Path::new(source_path);
    let destination_dir = Path::new(destination_dir);

    // Ensure the destination directory exists
    fs::create_dir_all(destination_dir)?;

    // Read the contents of the source file
    let mut file = File::open(source_path)?;
    let mut contents = Vec::new();
    file.read_to_end(&mut contents)?;

    // Compute a simple hash of the contents
    let mut hasher = DefaultHasher::new();
    hasher.write(&contents);
    let hash = hasher.finish();

    // Split the file name and extension, then reassemble with the hash
    let file_stem = source_path.file_stem().and_then(|name| name.to_str()).unwrap_or_default();
    let extension = source_path.extension().and_then(|ext| ext.to_str()).unwrap_or_default();
    let new_file_name = format!("{}-{:x}.{}", file_stem, hash, extension);

    let destination_path = destination_dir.join(new_file_name.clone());

    // Copy the file to the new destination
    fs::write(&destination_path, contents)?;

    Ok(new_file_name)
}
