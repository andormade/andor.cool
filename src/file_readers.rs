use crate::parsers::parse_content_with_front_matter;
use crate::types::{ContentCollection, ContentItem};
use std::fs;
use std::io::{Error, ErrorKind, Result};
use std::path::Path;

pub fn load_and_parse_file_with_front_matter(file_path: &Path) -> Result<ContentItem> {
    let content = fs::read_to_string(file_path).map_err(|e| {
        Error::new(
            e.kind(),
            format!("Failed to read file '{}': {}", file_path.display(), e),
        )
    })?;
    let mut parsed_content = parse_content_with_front_matter(&content);

    if let Some(file_stem) = file_path.file_stem().and_then(|s| s.to_str()) {
        parsed_content.insert("slug".to_string(), file_stem.to_string());
    }

    // Add file type to content for rendering pipeline
    if let Some(extension) = file_path.extension().and_then(|ext| ext.to_str()) {
        parsed_content.insert("file_type".to_string(), extension.to_string());
    }

    Ok(parsed_content)
}

pub fn load_and_parse_files_with_front_matter_in_directory(
    dir_path: &str,
) -> Result<ContentCollection> {
    let path = Path::new(dir_path);

    if !path.exists() {
        return Err(Error::new(
            ErrorKind::NotFound,
            format!("Directory '{}' does not exist. Make sure your site has the required directory structure.", dir_path),
        ));
    }

    let mut results = Vec::new();

    for entry in fs::read_dir(path).map_err(|e| {
        Error::new(
            e.kind(),
            format!("Failed to read directory '{}': {}", dir_path, e),
        )
    })? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(extension) = path.extension().and_then(|ext| ext.to_str()) {
                if extension == "md" || extension == "hbs" {
                    let parsed_content = load_and_parse_file_with_front_matter(&path)?;
                    results.push(parsed_content);
                }
            }
        }
    }

    results.sort_by(|a: &ContentItem, b| b["slug"].cmp(&a["slug"]));

    Ok(results)
}

pub fn load_site_config(site_name: &str) -> Result<ContentItem> {
    let config_path_str = format!("./sites/{site_name}/config.md");
    let config_path = Path::new(&config_path_str);
    if config_path.exists() {
        load_and_parse_file_with_front_matter(config_path)
    } else {
        // Return default configuration if no config file exists
        let mut default_config = ContentItem::new();
        default_config.insert("title".to_string(), String::new());
        default_config.insert("index_filename".to_string(), "index.html".to_string());
        Ok(default_config)
    }
}
