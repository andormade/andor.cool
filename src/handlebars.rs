use std::collections::HashMap;

pub fn replace_template_variable(template: &str, key: &str, value: &str) -> String {
    let mut result = template.to_string();

    let placeholder_with_space = format!("{{{{ {} }}}}", key);
    let placeholder_without_space = format!("{{{{{}}}}}", key);

    result = result.replace(&placeholder_with_space, value);
    result = result.replace(&placeholder_without_space, value);

    result
}

pub fn replace_template_variables(template: &str, variables: &HashMap<String, String>) -> String {
    let mut result = template.to_string();

    for (key, value) in variables {
        result = replace_template_variable(&result, key, value);
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_replace_template_variables() {
        let mut variables = HashMap::new();
        variables.insert("foo".to_string(), "apple".to_string());
        variables.insert("bar".to_string(), "banana".to_string());

        let template = "Lorem ipsum {{foo}} dolor {{bar}} sit amet.";
        let result = replace_template_variables(template, &variables);

        assert_eq!(result, "Lorem ipsum apple dolor banana sit amet.");
    }

    #[test]
    fn test_replace_template_variables_with_spaces() {
        let mut variables = HashMap::new();
        variables.insert("foo".to_string(), "apple".to_string());
        variables.insert("bar".to_string(), "banana".to_string());

        let template = "Lorem ipsum {{ foo }} dolor {{ bar }} sit amet.";
        let result = replace_template_variables(template, &variables);

        assert_eq!(result, "Lorem ipsum apple dolor banana sit amet.");
    }
}
