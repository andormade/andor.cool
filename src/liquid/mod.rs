pub mod _if;
pub mod include;
