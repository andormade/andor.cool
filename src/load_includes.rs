use std::collections::HashMap;
use std::fs;
use std::path::Path;
use crate::error::MyError;

pub fn load_liquid_includes(dir_path: &str) -> Result<HashMap<String, String>, MyError> {
    let path = Path::new(dir_path);
    let mut templates = HashMap::new();

    for entry in fs::read_dir(path)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_file() && path.extension().and_then(|ext| ext.to_str()) == Some("liquid") {
            if let Some(filename) = path.file_name().and_then(|name| name.to_str()) {
                let contents = fs::read_to_string(&path)?;
                templates.insert(filename.to_string(), contents);
            }
        }
    }

    Ok(templates)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use serde_json;
    use std::collections::BTreeMap;

    #[test]
    fn test_load_liquid_includes() {
        // Update test to handle Result
        let templates = load_liquid_includes("./_includes").unwrap_or_default();
        let sorted_templates: BTreeMap<_, _> = templates.into_iter().collect();
        let templates_json = serde_json::to_string_pretty(&sorted_templates).unwrap();
        // Use assert_snapshot! to compare the output against the stored snapshot
        assert_snapshot!(templates_json);
    }
}
