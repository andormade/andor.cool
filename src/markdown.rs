pub fn double_line_breaks_to_html(input: &str) -> String {
    input
        .split("\n\n")
        .map(|paragraph| format!("<p>{}</p>", paragraph))
        .collect::<Vec<_>>()
        .join("")
}

pub fn single_line_breaks_to_html(input: &str) -> String {
    input.replace("\n", "<br />")
}

pub fn list_to_html(input: &str) -> String {
    let mut result = String::new();
    let mut in_list = false;

    for line in input.lines() {
        if line.trim_start().starts_with('-') {
            if !in_list {
                in_list = true;
                result.push_str("<ul>");
            }
            result.push_str("<li>");
            result.push_str(&line.trim_start_matches('-').trim());
            result.push_str("</li>");
        } else {
            if in_list {
                in_list = false;
                result.push_str("</ul>");
            }
            result.push_str(line);
        }
    }

    if in_list {
        result.push_str("</ul>");
    }

    result
}

pub fn markdown_to_html(input: &str) -> String {
    //let mut html = double_line_breaks_to_html(input);
    let mut html = list_to_html(input);
    html = single_line_breaks_to_html(&html);
    html
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_list_to_html() {
        let markdown = r#"
- Item 1
- Item 2
- Item 3

Some other text.

- Item 1
- Item 2
"#;

        let expected_html = "<ul><li>Item 1</li><li>Item 2</li><li>Item 3</li></ul>Some other text.<ul><li>Item 1</li><li>Item 2</li></ul>";
        assert_eq!(list_to_html(markdown), expected_html);
    }

    #[test]
    fn test_double_line_breaks_to_html() {
        let markdown = r#"This is a line.
This is another line.

This is a new paragraph."#;
        let expected_html =
            "<p>This is a line.\nThis is another line.</p><p>This is a new paragraph.</p>";

        assert_eq!(double_line_breaks_to_html(markdown), expected_html);
    }

    #[test]
    fn test_single_line_breaks_to_html() {
        let markdown = r#"This is a line.
This is another line.

This is a new paragraph."#;
        let expected_html =
            "This is a line.<br />This is another line.<br /><br />This is a new paragraph.";

        assert_eq!(single_line_breaks_to_html(markdown), expected_html);
    }

    #[test]
    fn test_markdown_to_html_empty_input() {
        assert_eq!(markdown_to_html(""), "");
    }

    #[test]
    fn test_markdown_to_html_only_single_breaks() {
        let markdown = "Hello\nWorld";
        // list_to_html("Hello\nWorld") results in "HelloWorld" because it processes line by line
        // and concatenates them. No '\n' remains for single_line_breaks_to_html.
        let expected_html = "HelloWorld";
        assert_eq!(markdown_to_html(markdown), expected_html);
    }

    #[test]
    fn test_markdown_to_html_only_list() {
        let markdown = "- Item 1\n- Item 2";
        let expected_html = "<ul><li>Item 1</li><li>Item 2</li></ul>"; // single_line_breaks shouldn't affect this structure
        assert_eq!(markdown_to_html(markdown), expected_html);
    }

    #[test]
    fn test_markdown_to_html_list_with_spaces() {
        let markdown = "-   Item 1  \n- Item 2";
        let expected_html = "<ul><li>Item 1</li><li>Item 2</li></ul>";
        assert_eq!(markdown_to_html(markdown), expected_html);
    }

    #[test]
    fn test_markdown_to_html_list_with_internal_line_breaks() {
        // Current list_to_html processes line by line.
        // single_line_breaks_to_html will convert the \n within list items.
        // The variable 'markdown' here was unused and referred to a more complex case.
        // The actual test uses 'markdown_with_internal_break' defined later.
        // list_to_html output for a simple case: <ul><li>Item 1 still item 1</li><li>Item 2</li></ul> (assuming trim works well)
        // markdown_to_html output:
        // <ul><li>Item 1<br />  still item 1</li><li>Item 2</li></ul> (after single_line_breaks)
        // Let's trace:
        // list_to_html:
        // Line 1: "- Item 1" -> `result = "<ul><li>Item 1</li>"` (in_list = true)
        // Line 2: "  still item 1" -> `result = "<ul><li>Item 1</li>  still item 1"` (in_list = false, because it doesn't start with '-')
        // Line 3: "- Item 2" -> `result = "<ul><li>Item 1</li>  still item 1</ul><ul><li>Item 2</li>"` (in_list = true)
        // This is not what we want. The current list_to_html is basic.
        // The test `test_list_to_html` already shows how non-list lines are handled.
        //
        // If markdown is:
        // - Item 1 line 1
        // - Item 1 line 2
        // This becomes two separate list items.
        //
        // If markdown is:
        // - Item 1 part1\n  Item 1 part2 (this is not standard markdown list continuation)
        //   Our parser would treat "  Item 1 part2" as a separate line not part of the list.
        //
        // Let's test the actual behavior of the composition:
        // The variable 'markdown_input' here was unused.
        // list_to_html("- Item 1\nNon-list line\n- Item 2") = "<ul><li>Item 1</li></ul>Non-list line<ul><li>Item 2</li></ul>"
        // single_line_breaks_to_html applied to that: "<ul><li>Item 1</li></ul>Non-list line<br /><ul><li>Item 2</li></ul>"
        // This seems to be an error in my manual trace for `test_list_to_html` in the original code.
        // Let's re-verify `test_list_to_html`'s expected output.
        // Original test:
        // let markdown = "\n- Item 1\n- Item 2\n- Item 3\n\nSome other text.\n\n- Item 1\n- Item 2\n";
        // let expected_html = "<ul><li>Item 1</li><li>Item 2</li><li>Item 3</li></ul>Some other text.<ul><li>Item 1</li><li>Item 2</li></ul>";
        // This test implies that blank lines and non-list lines correctly terminate the list.
        // And `markdown_to_html` applies `single_line_breaks_to_html` to the output of `list_to_html`.
        //
        // So, for "list item with \n":
        let markdown_with_internal_break = "- Item 1\n  still part of item 1 conceptually, but not by current parser\n- Item 2";
        // list_to_html(markdown_with_internal_break):
        // 1. "- Item 1" -> `<ul><li>Item 1</li>`
        // 2. "  still part of item 1..." -> `</ul>  still part of item 1...` (list ends)
        // 3. "- Item 2" -> `<ul><li>Item 2</li></ul>` (new list starts)
        // So, list_to_html(markdown_with_internal_break) = "<ul><li>Item 1</li></ul>  still part of item 1 conceptually, but not by current parser<ul><li>Item 2</li></ul>"
        // Then, markdown_to_html applies single_line_breaks to this string. Since there are no '\n'
        // characters in the output of list_to_html for this input, single_line_breaks_to_html does nothing.
        let expected = "<ul><li>Item 1</li></ul>  still part of item 1 conceptually, but not by current parser<ul><li>Item 2</li></ul>";
        assert_eq!(markdown_to_html(markdown_with_internal_break), expected);
    }

    #[test]
    fn test_markdown_to_html_text_before_and_after_list() {
        let markdown = "Text before\n- Item 1\n- Item 2\nText after";
        // list_to_html: "Text before<ul><li>Item 1</li><li>Item 2</li></ul>Text after"
        // markdown_to_html: "Text before<br /><ul><li>Item 1</li><li>Item 2</li></ul>Text after"
        // No, list_to_html:
        // 1. "Text before" -> `Text before`
        // 2. "- Item 1" -> `<ul><li>Item 1</li>`
        // 3. "- Item 2" -> `<li>Item 2</li>`
        // 4. "Text after" -> `</ul>Text after`
        // So, list_to_html(markdown) = "Text before<ul><li>Item 1</li><li>Item 2</li></ul>Text after"
        // Then, markdown_to_html applies single_line_breaks:
        // "Text before<br /><ul><li>Item 1</li><li>Item 2</li></ul>Text after"
        // This needs careful checking of how `input.lines()` and `push_str(line)` interact.
        // If "Text before" is a line, then `result.push_str(line)` is called.
        // `list_to_html` logic:
        // line = "Text before" -> !in_list, result.push_str("Text before")
        // line = "- Item 1"    -> in_list=true, result.push_str("<ul><li>Item 1</li>")
        // line = "- Item 2"    -> result.push_str("<li>Item 2</li>")
        // line = "Text after"  -> in_list=false, result.push_str("</ul>Text after")
        // Correct. list_to_html output is "Text before<ul><li>Item 1</li><li>Item 2</li></ul>Text after"
        // No '\n' characters remain in this string for single_line_breaks_to_html to process.
        let expected_html = "Text before<ul><li>Item 1</li><li>Item 2</li></ul>Text after";
        assert_eq!(markdown_to_html(markdown), expected_html);
    }
}
