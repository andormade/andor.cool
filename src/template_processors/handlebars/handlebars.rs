use std::collections::HashMap;
use crate::error::{Error, Result};

/// Replaces a single Handlebars variable in a template with its value.
/// 
/// # Arguments
/// * `template` - The template string containing Handlebars variables
/// * `key` - The variable name to replace
/// * `value` - The value to replace the variable with
/// 
/// # Returns
/// * `Result<String>` - The template with the variable replaced or an error if malformed
pub fn replace_template_variable(template: &str, key: &str, value: &str) -> Result<String> {
    if !is_valid_variable_name(key) {
        return Err(Error::Handlebars(format!("Invalid variable name: {}", key)));
    }

    let mut result = String::with_capacity(template.len());
    let mut chars = template.chars().peekable();
    
    while let Some(current) = chars.next() {
        if current == '{' && chars.peek() == Some(&'{') {
            chars.next(); // Skip second '{'
            let mut var_name = String::new();
            
            // Skip whitespace
            while let Some(&c) = chars.peek() {
                if !c.is_whitespace() {
                    break;
                }
                chars.next();
            }
            
            // Collect variable name
            while let Some(&c) = chars.peek() {
                if c == '}' {
                    break;
                }
                var_name.push(chars.next().unwrap());
            }
            
            var_name = var_name.trim().to_string();
            
            // Check for closing braces
            if chars.next() != Some('}') || chars.next() != Some('}') {
                return Err(Error::Handlebars("Unclosed variable in template".to_string()));
            }
            
            if var_name == key {
                result.push_str(value);
            } else {
                result.push_str(&format!("{{{{ {} }}}}", var_name));
            }
        } else {
            result.push(current);
        }
    }
    
    Ok(result)
}

/// Replaces all Handlebars variables in a template with their corresponding values.
/// 
/// # Arguments
/// * `template` - The template string containing Handlebars variables
/// * `variables` - A HashMap containing variable names and their values
/// 
/// # Returns
/// * `Result<String>` - The template with all variables replaced or an error if malformed
pub fn replace_template_variables(template: &str, variables: &HashMap<String, String>) -> Result<String> {
    let mut result = template.to_string();
    
    for (key, value) in variables {
        result = replace_template_variable(&result, key, value)?;
    }
    
    Ok(result)
}

/// Validates if a variable name follows Handlebars naming conventions.
fn is_valid_variable_name(name: &str) -> bool {
    if name.is_empty() {
        return false;
    }
    
    let first_char = name.chars().next().unwrap();
    if !first_char.is_alphabetic() && first_char != '_' {
        return false;
    }
    
    name.chars().all(|c| c.is_alphanumeric() || c == '_' || c == '.')
}

/// Removes Handlebars variables from the input string.
/// This function will remove any content between {{ and }} including the braces.
/// 
/// # Arguments
/// * `input` - The input string containing Handlebars variables
/// 
/// # Returns
/// * `Result<String>` - The string with variables removed or an error if malformed
pub fn remove_handlebars_variables(input: &str) -> Result<String> {
    let mut result = String::with_capacity(input.len());
    let mut chars = input.chars().peekable();
    let mut in_variable = false;
    
    while let Some(current) = chars.next() {
        if current == '{' && chars.peek() == Some(&'{') {
            if in_variable {
                return Err(Error::Handlebars("Nested opening braces '{{' found inside a variable".to_string()));
            }
            in_variable = true;
            // Skip the second '{'
            chars.next();
            
            // Skip whitespace after '{{'
            while let Some(&c) = chars.peek() {
                if !c.is_whitespace() {
                    break;
                }
                chars.next();
            }
            continue;
        }
        
        if in_variable {
            if current == '}' && chars.peek() == Some(&'}') {
                in_variable = false;
                chars.next(); // Skip the second '}'
                continue;
            }
        } else {
            result.push(current);
        }
    }
    
    if in_variable {
        return Err(Error::Handlebars("Unclosed Handlebars variable".to_string()));
    }
    
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_replace_template_variables() {
        let mut variables = HashMap::new();
        variables.insert("foo".to_string(), "apple".to_string());
        variables.insert("bar".to_string(), "banana".to_string());

        let template = "Lorem ipsum {{foo}} dolor {{bar}} sit amet.";
        let result = replace_template_variables(template, &variables).unwrap();

        assert_eq!(result, "Lorem ipsum apple dolor banana sit amet.");
    }

    #[test]
    fn test_replace_template_variables_with_spaces() {
        let mut variables = HashMap::new();
        variables.insert("foo".to_string(), "apple".to_string());
        variables.insert("bar".to_string(), "banana".to_string());

        let template = "Lorem ipsum {{ foo }} dolor {{ bar }} sit amet.";
        let result = replace_template_variables(template, &variables).unwrap();

        assert_eq!(result, "Lorem ipsum apple dolor banana sit amet.");
    }

    #[test]
    fn test_remove_handlebars_variables() {
        let template = "Lorem ipsum {{foo}} dolor {{ bar }} sit amet.";
        let result = remove_handlebars_variables(template).unwrap();
        assert_eq!(result, "Lorem ipsum  dolor  sit amet.");
    }

    #[test]
    fn test_remove_handlebars_variables_with_error() {
        let template = "Lorem ipsum {{foo dolor {{ bar }} sit amet.";
        let err = remove_handlebars_variables(template).unwrap_err();
        assert!(matches!(err, Error::Handlebars(_)));
        if let Error::Handlebars(msg) = err {
            assert!(msg.contains("Nested opening braces"), "Error message should mention nested braces");
        }
    }

    #[test]
    fn test_nested_variables() {
        let template = "Hello {{user.name}} and {{deeply.nested.value}}!";
        let result = remove_handlebars_variables(template).unwrap();
        assert_eq!(result, "Hello  and !");
    }

    #[test]
    fn test_empty_template() {
        let template = "";
        let result = remove_handlebars_variables(template).unwrap();
        assert_eq!(result, "");
    }

    #[test]
    fn test_whitespace_handling() {
        let template = "Hello {{  spaced  }} world";
        let result = remove_handlebars_variables(template).unwrap();
        assert_eq!(result, "Hello  world");
    }

    #[test]
    fn test_unclosed_variable() {
        let template = "Hello {{name";
        let err = remove_handlebars_variables(template).unwrap_err();
        assert!(matches!(err, Error::Handlebars(_)));
        if let Error::Handlebars(msg) = err {
            assert!(msg.contains("Unclosed"), "Error message should mention unclosed variable");
        }
    }

    #[test]
    fn test_invalid_variable_name() {
        let template = "Hello {{123invalid}}";
        let result = replace_template_variable(template, "123invalid", "value");
        assert!(result.is_err());
    }

    #[test]
    fn test_is_valid_variable_name() {
        assert!(is_valid_variable_name("validName"));
        assert!(is_valid_variable_name("valid_name"));
        assert!(is_valid_variable_name("valid.nested.name"));
        assert!(!is_valid_variable_name(""));
        assert!(!is_valid_variable_name("123invalid"));
        assert!(!is_valid_variable_name("invalid!name"));
    }
}
