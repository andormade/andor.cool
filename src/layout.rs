use std::fs;
use crate::error::Result;
use std::path::Path;

use crate::handlebars::replace_template_variable;

pub fn load_layout(file: &str) -> Result<String> {
    let file_path = Path::new(file);
    let content = fs::read_to_string(file_path)?;
    Ok(content)
}

pub fn insert_body_into_layout(layout: &str, body: &str) -> Result<String> {
    replace_template_variable(layout, "body", body)
}
