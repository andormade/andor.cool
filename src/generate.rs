use crate::file_copier::copy_file_with_versioning;
use crate::file_readers::load_and_parse_markdown_files_with_front_matter_in_directory;
use crate::handlebars::remove_handlebars_variables;
use crate::handlebars::replace_template_variable;
use crate::handlebars::replace_template_variables;
use crate::layout::insert_body_into_layout;
use crate::layout::load_layout;
use crate::liquid::process_liquid_includes;
use crate::load_includes::load_liquid_includes;
use crate::markdown::markdown_to_html;
use crate::write::write_html_to_file;
use std::collections::HashMap;
use std::io::Result;

fn render_page(
    body: &str,
    directory: &str,
    slug: &str,
    layout: &str,
    includes: &HashMap<String, String>,
    variables: &HashMap<String, String>,
) -> Result<()> {
    let mut html = markdown_to_html(&body);
    let file_name = directory.to_string() + &slug + ".html";
    html = process_liquid_includes(&html, &includes);
    html = insert_body_into_layout(&layout, &html);
    html = replace_template_variables(&html, &variables);
    html = remove_handlebars_variables(&html);
    write_html_to_file(&file_name, &html)?;

    Ok(())
}

pub fn generate() -> Result<()> {
    let css_file_name = copy_file_with_versioning("./style.css", "./out/")?;
    let posts = load_and_parse_markdown_files_with_front_matter_in_directory("./_posts")?;
    let pages = load_and_parse_markdown_files_with_front_matter_in_directory("./_pages")?;
    let includes = load_liquid_includes("./_includes");

    let mut global_variables = HashMap::new();
    global_variables.insert(
        "title".to_string(),
        "Andor Polgar's Visual Journal".to_string(),
    );

    let main_layout_template = load_layout("./_layouts/main.html")?;
    let mut main_layout_variables = HashMap::new();
    main_layout_variables.insert("css_file_name".to_string(), css_file_name);
    let mut main_layout = replace_template_variables(&main_layout_template, &main_layout_variables);

    // Generate pagination pages
    let posts_per_page = 5;
    let post_chunks: Vec<Vec<HashMap<String, String>>> = posts
        .chunks(posts_per_page)
        .map(|chunk| chunk.to_vec())
        .collect();

    for (index, chunk) in post_chunks.iter().enumerate() {
        let mut html = String::new();
        html.push_str("<div class=\"postlist\">\n");
        for post in chunk {
            html.push_str(&replace_template_variables(
                &includes
                    .get("post.liquid")
                    .cloned()
                    .unwrap_or_else(String::new),
                &post,
            ));
        }
        html.push_str("</div>");
 
        let slug = if index == 0 {
            "index".to_string()
        } else {
            format!("page{}", index + 1)
        };

        render_page(
            &html,
            "out/",
            &slug,
            &main_layout,
            &includes,
            &global_variables,
        )?;
    }

    // Generate index page
    let list_item_template = includes
        .get("list_item.liquid")
        .cloned()
        .unwrap_or_else(String::new);
    let mut html_list = String::new();
    html_list.push_str("<ul class=\"postlist\">\n");
    for post in &posts {
        html_list.push_str(&replace_template_variables(&list_item_template, &post));
    }
    html_list.push_str("</ul>");
    let mut html = insert_body_into_layout(&main_layout, &html_list);
    html = replace_template_variable(&html, "title", "Andor Polgar's Visual Journal");
    html = remove_handlebars_variables(&html);
    write_html_to_file(&"out/index.html", &html)?;

    // Generate posts
    for post in &posts {
        global_variables.insert(
            "title".to_string(),
            post.get("title")
                .cloned()
                .unwrap_or_else(String::new)
                .to_owned()
                + " - "
                + "Andor Polgar's Visual Journal",
        );

        let slug = post.get("slug").cloned().unwrap_or_else(String::new);
        let pathname: String = "posts/".to_owned() + &slug;
        main_layout_variables.insert("pathname".to_string(), pathname);
        main_layout = replace_template_variables(&main_layout_template, &main_layout_variables);

        render_page(
            &post.get("content").map(|s| s.as_str()).unwrap_or(""),
            "out/posts/",
            &post.get("slug").map(|s| s.as_str()).unwrap_or(""),
            &main_layout,
            &includes,
            &global_variables,
        )?;
    }

    // Generate pages
    for page in &pages {
        global_variables.insert(
            "title".to_string(),
            page.get("title")
                .cloned()
                .unwrap_or_else(String::new)
                .to_owned()
                + " - "
                + "Andor Polgar's Visual Journal",
        );

        let slug = page.get("slug").cloned().unwrap_or_else(String::new);
        main_layout_variables.insert("pathname".to_string(), slug);
        main_layout = replace_template_variables(&main_layout_template, &main_layout_variables);

        render_page(
            &page.get("content").map(|s| s.as_str()).unwrap_or(""),
            "out/posts/",
            &page.get("slug").map(|s| s.as_str()).unwrap_or(""),
            &main_layout,
            &includes,
            &global_variables,
        )?;
    }

    Ok(())
}
