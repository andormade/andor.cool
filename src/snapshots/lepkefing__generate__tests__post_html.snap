---
source: src/generate.rs
expression: "read_file_content(\"out/posts/test-post.html\")"
---
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Test Post - Test Site</title>
    <link rel="stylesheet" href="/assets/style-84fc2985e8986ebd.css">
</head>
<body>
    <div class="content">
        <article>    <h1>Test Post</h1>    # Test PostThis is a test post content.<ul><li>Item 1</li><li>Item 2</li><li>Item 3</li></ul><div class="test-include">
    This is included content for testing.
</div>      <div class="test-include">
    This is included content for testing.
</div> </article> 
    </div>
    <footer>Test Site Footer</footer>
</body>
</html>
